src/DevAdventCalendarCompetition/DevAdventCalendarCompetition.Tests/UnitTests/Controllers/AdminControllerTests.cs

```diff
@@ -0,0 +1,218 @@
+
+            // Act
+            var result = controller.Index();
+
+            // Assert
+            var viewResult = Assert.IsType<ViewResult>(result);
+            var model = Assert.IsAssignableFrom<List<TestDto>>(viewResult.ViewData.Model);
+            Assert.Single(model);
+        }
+
+        [Fact]
+        public void AddTest_ModelStateIsInvalid_ReturnsViewWithInvalidModel()
+        {
+            // Arrange
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+            controller.ModelState.AddModelError("Number", "Required");
+
+            // Act
+            var result = controller.AddTest(GetTestViewModel());
+
+            // Assert
+            var viewResult = Assert.IsType<ViewResult>(result);
+            Assert.IsType<TestViewModel>(viewResult.ViewData.Model);
+        }
+
+        [Fact]
+        public void AddTest_TestExists_ReturnsViewWithError()
+        {
+            // Arrange
+            this._testServiceMock.Setup(x => x.GetTestByNumber(It.IsAny<int>())).Returns(GetTestDto());
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+
+            // Act
+            var result = controller.AddTest(GetTestViewModel());
+
+            // Assert
+            var allErrors = controller.ModelState.Values.SelectMany(v => v.Errors);
+            Assert.Single(allErrors);
+            Assert.Contains(allErrors, x => x.ErrorMessage == ExceptionsMessages.TestAlreadyExists);
+            var viewResult = Assert.IsType<ViewResult>(result);
+            Assert.IsType<TestViewModel>(viewResult.ViewData.Model);
+        }
+
+        [Fact]
+        public void AddTest_TestNotExists_ReturnsARedirectAndAddsTest()
+        {
+            // Arrange
+            this._testServiceMock.Setup(x => x.GetTestByNumber(It.IsAny<int>())).Returns((TestDto)null);
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+
+            // Act
+            var result = controller.AddTest(GetTestViewModel());
+
+            // Assert
+            this._adminServiceMock.Verify(x => x.AddTest(It.IsAny<TestDto>()), Times.Once);
+            var redirectToActionResult = Assert.IsType<RedirectToActionResult>(result);
+            Assert.Equal("Index", redirectToActionResult.ActionName);
+        }
+
+        [Fact]
+        public void StartTest_TestIsAlreadyStarted_ThrowsException()
+        {
+            // Arrange
+            var test = GetTest(TestStatus.Started);
+            this._adminServiceMock.Setup(x => x.GetTestById(test.Id)).Returns(test);
+            using var controller = new AdminController(this._adminServiceMock.Object, this._testServiceMock.Object);
+
+            // Act
+            Func<ActionResult> act = () => controller.StartTest(test.Id, "20");
```

Comment:
why do we need Func here?

Reply:
Just to stick to the arrange-act-confirm syntax, not by inlining it in the assertion section.
`var ex = Assert.Throws<InvalidOperationException>(() => controller.StartTest(test.Id, "20"););`